src/lib/draw-and-pick.js

```diff
@@ -53,91 +56,88 @@ export function pickLayers(gl, {
+    gl.blendEquation(gl.FUNC_ADD);
+
+    // Render all pickable layers
+    layers.forEach((layer, layerIndex) => {
       if (layer.props.visible && layer.props.pickable) {
 
-        // Clear the frame buffer, render and sample
-        gl.clear(GL.COLOR_BUFFER_BIT | GL.DEPTH_BUFFER_BIT);
-        const info = createInfo({
-          layer,
-          pixel: [x, y],
-          devicePixel: [deviceX, deviceY],
-          pixelRatio
-        });
+        // Encode layerIndex with alpha
+        gl.blendColor(0, 0, 0, (layerIndex + 1) / 255);
 
-        layer.pickLayer({
-          info,
-          uniforms: Object.assign({},
+        layer.drawLayer({
+          uniforms: Object.assign(
+            {renderPickingBuffer: 1, pickingEnabled: 1},
             layer.context.uniforms,
             getUniformsFromViewport(layer.context.viewport, layer.props),
             {layerIndex}
-          ),
-          pickEnableUniforms: {renderPickingBuffer: 1, pickingEnabled: 1},
-          pickDisableUniforms: {renderPickingBuffer: 0, pickingEnabled: 0},
-          deviceX, deviceY,
-          mode
+          )
         });
+      }
+    });
+
+    // Read color in the central pixel, to be mapped with picking colors
+    const color = new Uint8Array(4);
+    gl.readPixels(deviceX, deviceY, 1, 1, GL.RGBA, GL.UNSIGNED_BYTE, color);
+
+    // Decode alpha to layer index
+    const pickedLayer = layers[color[3] - 1];
+    const baseInfo = Object.assign(
+      createInfo([x, y], viewport),
+      {
+        devicePixel: [deviceX, deviceY],
+        pixelRatio
+      }
+    );
+
+    layers.forEach(layer => {
+      const info = Object.assign({}, baseInfo, {layer});
 
-        if (info.index >= 0) {
+      if (layer === pickedLayer) {
+        const index = layer.decodePickingColor(color);
+        info.index = index;
+        info.color = color;
+        if (index >= 0) {
           info.picked = true;
-          info.zOrder = zOrder++;
           // If props.data is an indexable array, get the object
           if (Array.isArray(layer.props.data)) {
-            info.object = layer.props.data[info.index];
+            info.object = layer.props.data[index];
           }
         }
-
-        pickedInfos.push(info);
       }
-    }
-  });
 
-  // Calling callbacks can have async interactions with React
-  // which nullifies layer.state.
-  const unhandledPickInfos = [];
-  for (const info of pickedInfos) {
-    let handled = null;
-    switch (mode) {
-    case 'click': handled = info.layer.props.onClick(info); break;
-    case 'hover': handled = info.layer.props.onHover(info); break;
-    default: throw new Error('unknown pick type');
-    }
+      const handled = layer.pickLayer({info, mode});
 
-    if (!handled) {
-      unhandledPickInfos.push(info);
-    }
-  }
+      if (!handled) {
+        unhandledPickInfos.push(info);
+      }
+    });
+
+    // restore blend mode
+    setBlendMode(gl, oldBlendMode);
```

Comment:
@shaojingli Note that an exception during picking leaves us in a bad state here. The proposed new state management in luma.gl would also handle this case.